src/native/libs/System.Native/pal_interfaceaddresses.c

```diff
@@ -95,14 +100,75 @@ static inline uint8_t mask2prefix(uint8_t* mask, int length)
 
     return len;
 }
+
+#if !HAVE_IFADDRS && TARGET_ANDROID
+// This structure is exactly the same as struct ifaddrs defined in ifaddrs.h but since the header
+// might not be available (e.g., in bionics used in Android before API 24) we need to mirror it here
+// so that we can dynamically load the getifaddrs function and use it.
+struct ifaddrs
+{
+	struct ifaddrs *ifa_next;
+	char *ifa_name;
+	unsigned int ifa_flags;
+	struct sockaddr *ifa_addr;
+	struct sockaddr *ifa_netmask;
+	union
+	{
+		struct sockaddr *ifu_broadaddr;
+		struct sockaddr *ifu_dstaddr;
+	} ifa_ifu;
+	void *ifa_data;
+};
+#endif
+
+#if !HAVE_GETIFADDRS && TARGET_ANDROID
+// Try to load the getifaddrs and freeifaddrs functions manually.
+// This workaround is necessary on Android prior to API 24 and it can be removed once
+// we drop support for earlier Android versions.
+static int (*getifaddrs)(struct ifaddrs**) = NULL;
+static void (*freeifaddrs)(struct ifaddrs*) = NULL;
+
+static bool ensure_getifaddrs_is_loaded()
+{
+    static bool loading_already_attempted = false;
+    static pthread_mutex_t lock = PTHREAD_MUTEX_INITIALIZER;
+
+    if (pthread_mutex_lock(&lock) == 0)
```

Comment:
Maybe consider using atomics here? Either the `__atomic` intrinsics (supported by clang and gcc) or, if the sources are built with C11, the standard atomic library (`stdatomic.h`). Mutex is a bit heavy and shouldn't be necessary here.

Reply:
I replaced mutexes with atomic operations. Please check out the updated implementation.

Reply:
It looks good :)